### PR TITLE
Add Subversion support

### PR DESCRIPTION
**Pull request by tsimonq2**
This adds support for Subversion and fixes [bug 1543243](https://bugs.launchpad.net/snapcraft/+bug/1543243) in Launchpad.

### CODE REVIEW

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by elopio** [comment]
Thanks for your contribution @tsimonq2, and welcome to snapcraft!
I'll leave a note here for the rest that you have already signed the contributors agreement. Thanks for that too.

This branch is missing unit tests. Take a look at snapcraft/tests/test_sources.py for the kind of things that we require for full code coverage. Let me know if you need a hand with the tests.

File: integration_tests/snaps/simple-svn/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+version: 0.1
+summary: svn integration test
+description: use svn
+confinement: strict
+
+parts:
+    svn:
+        plugin: make
+        source: .
+        source-type: svn
```
Comment (elopio) [inline integration_tests/snaps/simple-svn/snapcraft.yaml]: This looks good, but it's missing that make file, and the test file. I'm not sure if it's work in progress, or you forgot to add the files to git.

File: snapcraft/internal/sources.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#-*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
```
Comment (elopio) [inline snapcraft/internal/sources.py]: Not a big deal, but please revert this change.

File: snapcraft/internal/sources.py [hunk 3]
```diff
@@ -383,6 +403,8 @@ def _get_source_type_from_uri(source, ignore_errors=False):
         source_type = 'tar'
     elif source.endswith('.zip'):
         source_type = 'zip'
+    elif source.startswith('svn:'):
+        source_type = 'subversion'
```
Comment (elopio) [inline snapcraft/internal/sources.py]: another nit, please move this case after the git one. I'm a little code order freak, and I would like that all the cvs sources are together.

File: snapcraft/internal/sources.py (hunk 2 above)
Comment (tsimonq2) [inline snapcraft/internal/sources.py]: It was an accident, I'll revert it :)

File: integration_tests/snaps/simple-svn/snapcraft.yaml (hunk 1 above)
Comment (chadmiller) [inline integration_tests/snaps/simple-svn/snapcraft.yaml]: It would be nice to test "svn:" source lines, too, and one check-out and one update.

Comment (elopio) [inline integration_tests/snaps/simple-svn/snapcraft.yaml]: woohoo, Chad is here! :tada:

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

File: integration_tests/snaps/simple-svn/snapcraft.yaml (hunk 1 above)
Comment (tsimonq2) [inline integration_tests/snaps/simple-svn/snapcraft.yaml]: @elopio - I'm a little confused here. the git-* and hg-* directories have almost the same contents. What is different, and if applicable, what should I put in the make file and test file?

@chadmiller - Sort of going with the comment I left for elopio, would you suggest I do it as separate parts in this snapcraft.yaml, or in separate tests, and if so, would I rename the directory from simple-svn to svn-*?

File: snapcraft/internal/sources.py [hunk 4]
```diff
@@ -27,7 +27,7 @@
     directory tree or a tarball or a revision control repository
     ('git:...').
 
-  - source-type: git, bzr, hg, tar or zip
+  - source-type: git, bzr, hg, tar, zip, or subversion
```
Comment (kyrofa) [inline snapcraft/internal/sources.py]: I know it's stupid, but can we have svn next to the other version control systems and leave tar and zip at the end?

File: snapcraft/internal/sources.py [hunk 5]
```diff
@@ -359,6 +378,7 @@ def get_required_packages(options):
     'mercurial': Mercurial,
     'tar': Tar,
     'zip': Zip,
+    'subversion': Subversion,
```
Comment (kyrofa) [inline snapcraft/internal/sources.py]: Is there a reason not to also resolve 'svn' to `Subversion`?

File: integration_tests/snaps/simple-svn/snapcraft.yaml (hunk 1 above)
Comment (kyrofa) [inline integration_tests/snaps/simple-svn/snapcraft.yaml]: Unless I'm mistaken, it doesn't look like `source-type: svn` will actually work given the code (i.e. it would need to be `source-type: subversion`)

File: snapcraft/internal/sources.py (hunk 5 above)
Comment (tsimonq2) [inline snapcraft/internal/sources.py]: Good point, nice catch, @elopio

File: integration_tests/snaps/simple-svn/snapcraft.yaml (hunk 1 above)
Comment (tsimonq2) [inline integration_tests/snaps/simple-svn/snapcraft.yaml]: @kyrofa Fixed in the recent set of commits I pushed

File: snapcraft/tests/test_sources.py [hunk 6]
```diff
@@ -321,6 +321,22 @@ def test_init_with_source_branch_and_tag_raises_exception(self):
...
+        svn.pull()
+        self.mock_run.assert_called_once_with(
+            ['svn', 'checkout', 'svn://my-source', 'source-dir'])
+
+    def test_pull_existing(self):
+        self.mock_path_exists.return_value = True
+        svn = sources.Subversion('svn://my-source', 'source_dir')
+        svn.pull()
+        self.mock_run.assert_called_once_with(
+            ['svn', 'update', 'source-dir'])
```
Comment (tsimonq2) [inline snapcraft/tests/test_sources.py]: @elopio This line is causing something to fail, I think the unit tests? https://travis-ci.org/ubuntu-core/snapcraft/jobs/137352934

What do I need to correct to make this go away? Do I need to remove `'svn://my-source'` from line 334?

File: integration_tests/snaps/simple-svn/snapcraft.yaml (hunk 1 above)
Comment (chadmiller) [inline integration_tests/snaps/simple-svn/snapcraft.yaml]: @tsimonq2, I also think it makes us worry that you didn't run the test and discover the problem, which is kind of bad; or that you did run tests and this test should have run and failed but did not, which is really really bad. Please find out what the problem is, here.

Comment (tsimonq2) [inline integration_tests/snaps/simple-svn/snapcraft.yaml]: @chadmiller If you are talking about kyrofa's issue, I found that out already, but I forgot to update my code. My bad, but it's fixed now.

What else are you referring to if not that? Can you clarify?

Comment (chadmiller) [inline integration_tests/snaps/simple-svn/snapcraft.yaml]: @tsimonq2, yes, I mean that the source-type couldn't have matched anything in the test, earlier, and should have caused a test failure.

Comment (tsimonq2) [inline integration_tests/snaps/simple-svn/snapcraft.yaml]: @chadmiller Well like I said, I figured that out myself but I forgot to edit the code. It was a simple mistake that I'm glad was caught, and I'm thankful for that. That's all.

Now, what do I do now with this directory and this snapcraft.yaml file? See comment #4 on this line comment string(if that makes any sense).

File: snapcraft/tests/test_sources.py (hunk 6 above)
Comment (sergiusens) [inline snapcraft/tests/test_sources.py]: on 334 you specify `source_dir` and on 337 you expect `source-dir`

Comment (tsimonq2) [inline snapcraft/tests/test_sources.py]: @sergiusens oh! nice catch, thanks.

File: snapcraft/internal/sources.py [hunk 7]
```diff
@@ -292,6 +292,27 @@ def provision(self, dst, clean_target=True, keep_zip=False):
...
+        if source_tag:
+            raise IncompatibleOptionsError(
+                'can\'t specify a source-tag for a subversion source')
+        elif source_branch:
+            raise IncompatibleOptionsError(
+                'can\'t specify a source-branch for a subversion source')
+
+    def pull(self):
+        if os.path.exists(os.path.join(self.source_dir, '.svn')):
+            subprocess.check_call(
```
Comment (sergiusens) [inline snapcraft/internal/sources.py]: why not

```python
subprocess.check_call(['svn', 'update'], cwd=self.source_dir)
```
?

File: snapcraft/internal/sources.py [hunk 8]
```diff
@@ -348,6 +369,8 @@ def get_required_packages(options):
         packages.append('tar')
     elif source_type == 'hg' or source_type == 'mercurial':
         packages.append('mercurial')
+    elif source_type == 'subversion':
```
Comment (sergiusens) [inline snapcraft/internal/sources.py]: this should be
```python
elif source_type == 'subversion' or source_type == 'svn'
```

File: snapcraft/tests/test_sources.py [hunk 9]
```diff
@@ -321,6 +321,22 @@ def test_init_with_source_branch_and_tag_raises_exception(self):
...
+        svn.pull()
+        self.mock_run.assert_called_once_with(
+            ['svn', 'checkout', 'svn://my-source', 'source_dir'])
+
+    def test_pull_existing(self):
+        self.mock_path_exists.return_value = True
+        svn = sources.Subversion('svn://my-source', 'source_dir')
+        svn.pull()
+        self.mock_run.assert_called_once_with(
+            ['svn', 'update', 'source_dir'])
```
Comment (sergiusens) [inline snapcraft/tests/test_sources.py]: why not

```python
self.mock_run.assert_called_once_with(['svn', 'update'], cwd=self.source_dir)
```
 if you take the advice from above

File: snapcraft/internal/sources.py (hunk 7 above)
Comment (tsimonq2) [inline snapcraft/internal/sources.py]: Good call, I'll make this change

File: snapcraft/internal/sources.py (hunk 8 above)
Comment (tsimonq2) [inline snapcraft/internal/sources.py]: Nice catch, I'm glad this was caught, I'll make the change

File: snapcraft/tests/test_sources.py (hunk 9 above)
Comment (tsimonq2) [inline snapcraft/tests/test_sources.py]: Yeah, I prefer this over the code I had. Thanks!

**Comment by sergiusens** [comment]
For a proper integration test you will need to create this layout either on the fly or commited (on the fly might be more versatile)

Here's a shell snippet

```
svncreate repo
svn checkout file:///$(pwd)/repo checkout1 #  the source entry in the test code would use this
svn checkout file:///$(pwd)/repo checkout2
cd checkout2
touch some_file
svn add file
svn commit -m 'some file'
cd ../checkout1
svn update # this would be running pull again for the source
ls some_file
```

**Comment by sergiusens** [comment]
@elopio do you mind looking at that integration test I mention ^^

File: snapcraft/sources.py [hunk 10]
```diff
@@ -20,4 +20,5 @@
 from snapcraft.internal.sources import Tar         # noqa
 from snapcraft.internal.sources import Local       # noqa
 from snapcraft.internal.sources import Zip         # noqa
+from snapcraft.internal.sources import Subversion
```
Comment (sergiusens) [inline snapcraft/sources.py]: this is missing a ` # noqa` statement, line it up with the `Zip` one above

Comment (tsimonq2) [inline snapcraft/sources.py]: All fixed, thanks :)

File: integration_tests/test_svn_source.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2015 Canonical Ltd
```
Comment (elopio) [inline integration_tests/test_svn_source.py]: 2016

File: integration_tests/test_svn_source.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+            ['cd', 'local'],stdout=subprocess.DEVNULL)
+        subprocess.check_call(
+            ['echo', 'test', '>', 'test_file'],stdout=subprocess.DEVNULL)
+        subprocess.check_call(
+            ['svn', 'commit', '-m', 'test2'],stdout=subprocess.DEVNULL)
+
+        subprocess.check_call(
+            ['cd', '../testupdate'],stdout=subprocess.DEVNULL)
+        subprocess.check_call(
+            ['svn', 'update', '.'],stdout=subprocess.DEVNULL)
```
Comment (elopio) [inline integration_tests/test_svn_source.py]: As I discussed on irc, this is missing the calls to self.run_snapcraft('pull', project_dir)

File: snapcraft/internal/sources.py [hunk 13]
```diff
@@ -201,6 +201,27 @@ def pull(self):
...
 
 
+class Subversion(Base):
+
+    def __init__(self, source, source_dir, source_tag=None,
+                 source_branch=None):
+        super().__init__(source, source_dir, source_tag, source_branch)
+        if source_tag:
+            raise IncompatibleOptionsError(
+                'can\'t specify a source-tag for a subversion source')
```
Comment (elopio) [inline snapcraft/internal/sources.py]: When the string has a single quote, we usually write it with double quotes:
"can't specify a source-tag..."

File: snapcraft/internal/sources.py [hunk 14]
```diff
@@ -201,6 +201,27 @@ def pull(self):
...
+        elif source_branch:
+            raise IncompatibleOptionsError(
+                'can\'t specify a source-branch for a subversion source')
```
Comment (elopio) [inline snapcraft/internal/sources.py]: ditto

File: snapcraft/tests/test_sources.py [hunk 15]
```diff
@@ -321,6 +321,22 @@ def test_init_with_source_branch_and_tag_raises_exception(self):
...
+            ['svn', 'checkout', 'svn://my-source', 'source_dir'])
+
+    def test_pull_existing(self):
+        self.mock_path_exists.return_value = True
+        svn = sources.Subversion('svn://my-source', 'source_dir')
+        svn.pull()
+        self.mock_run.assert_called_once_with(
+            ['svn', 'update'], cwd=svn.source_dir)
+
+
```
Comment (elopio) [inline snapcraft/tests/test_sources.py]: you are missing two tests to check the errors when tag and branch are used.

File: snapcraft/sources.py [hunk 16]
```diff
@@ -17,6 +17,7 @@
 from snapcraft.internal.sources import Bazaar      # noqa
 from snapcraft.internal.sources import Git         # noqa
 from snapcraft.internal.sources import Mercurial   # noqa
+from snapcraft.internal.sources import Subversion  # noqa
 from snapcraft.internal.sources import Tar         # noqa
 from snapcraft.internal.sources import Local       # noqa
 from snapcraft.internal.sources import Zip         # noqa
```
Comment (elopio) [inline snapcraft/sources.py]: Can you please update the copyright year in this file to 2015, 2016 ?

Comment (tsimonq2) [inline snapcraft/sources.py]: @elopio Sure

File: snapcraft/internal/sources.py [hunk 17]
```diff
@@ -201,6 +201,27 @@ def pull(self):
...
 
 
+class Subversion(Base):
+
+    def __init__(self, source, source_dir, source_tag=None,
+                 source_branch=None):
+        super().__init__(source, source_dir, source_tag, source_branch)
+        if source_tag:
+            raise IncompatibleOptionsError(
+                "can\'t specify a source-tag for a subversion source")
```
Comment (elopio) [inline snapcraft/internal/sources.py]: you can remove the \

Comment (tsimonq2) [inline snapcraft/internal/sources.py]: @elopio Bazaar has the same problem, I'll correct Subversion, but might it be worth a whole other PR to fix this?

File: snapcraft/internal/sources.py [hunk 18]
```diff
@@ -201,6 +201,27 @@ def pull(self):
         subprocess.check_call(cmd)
 
 
+class Subversion(Base):
+
+    def __init__(self, source, source_dir, source_tag=None,
+                 source_branch=None):
+        super().__init__(source, source_dir, source_tag, source_branch)
+        if source_tag:
+            raise IncompatibleOptionsError(
```
Comment (tsimonq2) [inline snapcraft/internal/sources.py]: @elopio this is causing the coverage decrease of 0.03%

File: snapcraft/internal/sources.py [hunk 19]
```diff
@@ -201,6 +201,27 @@ def pull(self):
...
+                "can't specify a source-tag for a Subversion source")
+        elif source_branch:
+            raise IncompatibleOptionsError(
```
Comment (tsimonq2) [inline snapcraft/internal/sources.py]: ditto

File: integration_tests/test_svn_source.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+        os.chdir(project_dir)
+
+        self._init_svn()
+
+        subprocess.check_call(
+            ['svn', 'checkout', 'file:///$(pwd)/repo', 'local'],
+            stdout=subprocess.DEVNULL)
+
+        os.chdir("local/")
+        subprocess.check_call(['touch', 'file'],stdout=subprocess.DEVNULL)
```
Comment (elopio) [inline integration_tests/test_svn_source.py]: Generally we call subprocess when there is no good way to do it with pure python. In the case of touch, you can do instead:

open(os.path.join('local', 'file'), 'w').close()

File: integration_tests/test_svn_source.py [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+        subprocess.check_call(
+~           ['svn', 'commit', '-m', 'test'],stdout=subprocess.DEVNULL)
+        subprocess.check_call(['svn', 'update'],stdout=subprocess.DEVNULL)
+        subprocess.check_call(
+            ['rm', '-rf', 'local/'],stdout=subprocess.DEVNULL)
+        os.chdir("..")
+
+        self.run_snapcraft('pull', project_dir)
+        revno = subprocess.check_output('svnversion parts/svn/src').strip()
+        self.assertEqual('"1"', revno)
```
Comment (elopio) [inline integration_tests/test_svn_source.py]: You can also check that the file exists.

Comment (tsimonq2) [inline integration_tests/test_svn_source.py]: @elopio Good call

Comment (tsimonq2) [inline integration_tests/test_svn_source.py]: Done

File: snapcraft/sources.py (hunk 16 above)
Comment (tsimonq2) [inline snapcraft/sources.py]: Done

**Comment by tsimonq2** [comment]
@sergiusens Integrated ;)

**Comment by tsimonq2** [comment]
This seems to me like it's close to done if not completely done given the feedback I've received. Does anybody notice anything else that I should correct?

File: integration_tests/test_svn_source.py [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+        super().setUp()
+        if shutil.which('svn') is None:
+            self.skipTest('svn is not installed')
+
+    def _init_svn(self):
+        subprocess.check_call(
+            ['svnadmin', 'create', 'repo'], stdout=subprocess.DEVNULL)
+
+    def test_pull_svn(self):
+
```
Comment (elopio) [inline integration_tests/test_svn_source.py]: a nit, remove this empty line.

File: integration_tests/test_svn_source.py [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+            ['svn', 'add', 'file'],stdout=subprocess.DEVNULL)
+        subprocess.check_call(
+~           ['svn', 'commit', '-m', 'test'],stdout=subprocess.DEVNULL)
+        subprocess.check_call(['svn', 'update'],stdout=subprocess.DEVNULL)
+        subprocess.check_call(
+            ['rm', '-rf', 'local/'],stdout=subprocess.DEVNULL)
+        os.chdir("..")
+
+        self.run_snapcraft('pull', project_dir)
+        revno = subprocess.check_output('svnversion parts/svn/src').strip()
```
Comment (elopio) [inline integration_tests/test_svn_source.py]: please use os.path.join('parts', 'svn', 'src')

File: integration_tests/test_svn_source.py [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+        self.assertEqual('"1"', revno)
+        filepresent = os.path.isfile("parts/svn/src/file")
+        self.assertEqual('"True"', filepresent)
```
Comment (elopio) [inline integration_tests/test_svn_source.py]: You can use self.assertThat(os.path.join(parts_src_path, 'file'), FileExists())
Add to the top:
from testtools.matchers import FileExists. Between shutil and integration_tests, with empty lines around.

**Comment by elopio** [comment]
Looks good to me. Please rebase or merge with master, or hit the update button that git shows you here vvv.

File: integration_tests/test_svn_source.py [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+        os.chdir("local/")
+        open(os.path.join('local', 'file'), 'w').close()
+        subprocess.check_call(
+            ['svn', 'add', 'file'],stdout=subprocess.DEVNULL)
+        subprocess.check_call(
+~           ['svn', 'commit', '-m', 'test'],stdout=subprocess.DEVNULL)
+        subprocess.check_call(['svn', 'update'],stdout=subprocess.DEVNULL)
+        subprocess.check_call(
+            ['rm', '-rf', 'local/'],stdout=subprocess.DEVNULL)
+        os.chdir("..")
```
Comment (chadmiller) [inline integration_tests/test_svn_source.py]: The 'chdir's bother me. Use instead the "cwd" keyword on subprocess functions. Or at least try/finally to save and restore the current working dir for the process.

Comment (tsimonq2) [inline integration_tests/test_svn_source.py]: @chadmiller can you clarify what you mean by this?

File: integration_tests/test_svn_source.py [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+
+    def test_pull_svn(self):
+
+        project_dir = self.copy_project_to_tmp('svn-checkout')
+        os.chdir(project_dir)
+
+        self._init_svn()
+
+        subprocess.check_call(
+            ['svn', 'checkout', 'file:///$(pwd)/repo', 'local'],
```
Comment (chadmiller) [inline integration_tests/test_svn_source.py]: That $() is not going to do what you think unless subversion is going to interpret it, which I doubt. Please verify.

File: integration_tests/test_svn_source.py (hunk 25 above)
Comment (chadmiller) [inline integration_tests/test_svn_source.py]: subprocess.check_call( ['svn', 'add', 'file'],stdout=subprocess.DEVNULL, pwd=project_dir)

not

chdir    
s.check_call

If this is a standalone test program, this doesn't matter very much, but changing global state in a test that continues past your test is a bad idea.

File: integration_tests/test_svn_source.py (hunk 26 above)
Comment (tsimonq2) [inline integration_tests/test_svn_source.py]: @chadmiller running that command works locally, and playing around with it, I'm not seeing another way to do it. I think it's fine.

Comment (elopio) [inline integration_tests/test_svn_source.py]: 'file://{}'.format(os.path.join(project_dir, 'repo'))

Comment (tsimonq2) [inline integration_tests/test_svn_source.py]: @elopio thanks, didn't know that was something I could do!

**Comment by tsimonq2** [comment]
@chadmiller Does it look good now?

@elopio I've merged with master, it should be good to go now.

**Comment by elopio** [comment]
One more thing I forgot: please add subversion to debian/tests/control, on the integration section.

File: integration_tests/test_svn_source.py [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+
+        self._init_svn()
+
+        subprocess.check_call(
+            ['svn', 'checkout', 'file:///{}'.format(os.path.join(project_dir, 'repo')), 'local'],
+            stdout=subprocess.DEVNULL)
+
+        open(os.path.join('local', 'file'), 'w').close()
+        subprocess.check_call(
+            ['svn', 'add', 'file'],stdout=subprocess.DEVNULL,cwd='local/')
```
Comment (elopio) [inline integration_tests/test_svn_source.py]: This will fail the pep8 check. You need to put one empty space after the commas.

**Comment by elopio** [comment]
Looking at the coveralls coverage, you are still missing one test for the uri.
Take a look in test_sources.py, the TestUri test case. You need to add one that's just like the git test.

**Comment by tsimonq2** [comment]
@elopio The only other thing I can think to fix is https://coveralls.io/builds/6601979/source?filename=snapcraft%2Finternal%2Fsources.py#L378

What file would I edit to add coverage for this?

**Comment by elopio** [comment]
Look at snapcraft/tests/test_yaml.py, test_config_adds_vcs_packages...

**Comment by tsimonq2** [comment]
@elopio thanks

**Comment by elopio** [comment]
OK to test

**Comment by elopio** [comment]
retest this please

File: snapcraft/internal/sources.py [hunk 28]
```diff
@@ -201,6 +201,39 @@ def pull(self):
...
+
+    def pull(self):
+        if os.path.exists(os.path.join(self.source_dir, '.svn')):
+            subprocess.check_call(
+                ['svn', 'update'], cwd=self.source_dir)
+        else:
+            if self.source.startswith('svn:') or \
+                    self.source.startswith('https:') or \
+                    self.source.startswith('http:') or \
+                    self.source.startswith('svn+ssh:'):
```
Comment (elopio) [inline snapcraft/internal/sources.py]: I'm wondering what would be a better way to do this.
what about reversing the loging and
if os.path.isdir(self.source), append file://, else, leave it alone ?

Comment (tsimonq2) [inline snapcraft/internal/sources.py]: @elopio I'll play around with this.

Comment (tsimonq2) [inline snapcraft/internal/sources.py]: @elopio that's a better solution, thanks, pushing now

File: snapcraft/tests/test_sources.py [hunk 29]
```diff
@@ -321,6 +321,58 @@ def test_init_with_source_branch_and_tag_raises_exception(self):
...
+            ['svn', 'checkout', 'svn://my-source', 'source_dir'])
+
+    def test_pull_local(self):
+        svn = sources.Subversion(
+            'file://{}'.format(os.path.abspath('my-source/')), 'source_dir')
+        svn.pull()
+        self.mock_run.assert_called_once_with(
+            ['svn', 'checkout',
+             'file://{}'.format(os.path.abspath('my-source/')),
+             'source_dir'])
```
Comment (elopio) [inline snapcraft/tests/test_sources.py]: notice that here you are not testing anything, because you are passing to the Subversion constructor a source that will return false in isdir.
On the constructor, pass only 'my-source'.

File: integration_tests/test_svn_pull_update.py [hunk 30]
```diff
@@ -0,0 +1,77 @@
...
+    def setUp(self):
+        super().setUp()
+        if shutil.which('svn') is None:
+            self.skipTest('svn is not installed')
+
+    def _init_svn(self):
+        subprocess.check_call(
+            ['svnadmin', 'create', 'repo'], stdout=subprocess.DEVNULL)
+
+    def test_pull_svn(self):
```
Comment (elopio) [inline integration_tests/test_svn_pull_update.py]: You can put this test in the same file as the other. Call one test_pull_svn_checkout and the other test_pull_svn_update.

Comment (tsimonq2) [inline integration_tests/test_svn_pull_update.py]: Oh okay, I'll do this, thanks

**Comment by tsimonq2** [comment]
retest this please

**Comment by tsimonq2** [comment]
Seems like the autopkgtest failure is not related to my snap if I'm reading this correctly: http://paste.ubuntu.com/17582500/

**Comment by tsimonq2** [comment]
Nope, found it

File: integration_tests/test_svn_pull.py [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+        subprocess.check_call(
+            ['svn', 'commit', '-m', 'test'], stdout=subprocess.DEVNULL, cwd='local/')
+        subprocess.check_call(['svn', 'update'], stdout=subprocess.DEVNULL, cwd='local/')
+        subprocess.check_call(
+            ['rm', '-rf', 'local/'], stdout=subprocess.DEVNULL)
+
+        part_src_path = os.path.join('parts', 'svn', 'src')
+        self.run_snapcraft('pull', project_dir)
+        revno = subprocess.check_output(['svnversion', part_src_path]).strip()
+        self.assertEqual(1, revno)
```
Comment (tsimonq2) [inline integration_tests/test_svn_pull.py]: This is causing the integration test errors. I have no clue what I can do to fix this. Thoughts? @elopio

Comment (elopio) [inline integration_tests/test_svn_pull.py]: Hey, sorry for the delay.
check_output never returns an int. By default, it returns bytes. I think it's better if you pass universal_newlines=True, and then assertEqual('1', revno)

Comment (tsimonq2) [inline integration_tests/test_svn_pull.py]: thanks @elopio
